compiler/src/dotty/tools/dotc/ast/TreeInfo.scala

```diff
@@ -629,6 +629,28 @@ trait TypedTreeInfo extends TreeInfo[Type] { self: Trees.Instance[Type] =>
       case nil =>
         Nil
     }
+
+  /** Is this a selection of a member of a structural type that is not a member
+   *  of an underlying class or trait?
+   */
+  def isStructuralTermSelect(tree: Tree)(implicit ctx: Context) = tree match {
+    case tree: Select =>
+      def hasRefinement(qualtpe: Type): Boolean = qualtpe.dealias match {
+        case RefinedType(parent, rname, rinfo) =>
+          rname == tree.name && tree.tpe.widen <:< rinfo || hasRefinement(parent)
+        case tp: TypeProxy =>
+          hasRefinement(tp.underlying)
+        case tp: OrType =>
+          hasRefinement(tp.tp1) || hasRefinement(tp.tp2)
```

Comment:
Ah I see! The `AndType` case makes sense, the `OrType` case seems suspicious but I don't think it matters since we no longer allow calling a method on a union type that matches more than one symbol, e.g. the following doesn't compile:
```scala
class Closeable {
  def close(): Unit = {}
}


object Test {
  import scala.reflect.Selectable.reflectiveSelectable

  def f(closeable: (Selectable { def close(): Unit }) | Closeable) =
    closeable.close() // error: value `close` is not a member of (Selectable{close: ()Unit} | Closeable)(closeable)

  def main(args: Array[String]): Unit = {
    f(new Closeable)
  }
}
```